File: 2022/07/0710/미로탐색.py
from collections import deque
from heapq import heappop, heappush
from pprint import pprint
import sys
sys.stdin = open('./text/2178.txt', 'r')
input = sys.stdin.readline
def Split():return map(int, input().strip().split())
def Int():return int(input().strip())

Y, X = Split()
board = [list(input().strip()) for _ in range(Y)]
visited = [[0] * X for _ in range(Y)]
di = ((-1, 0), (1, 0), (0, 1), (0, -1))
Q = deque([(0, 0)])
visited[0][0] = 1
while Q:
    y, x = Q.popleft()
    for dy, dx in di:
        ny = y + dy
        nx = x + dx
        if 0 <= ny < Y and 0 <= nx < X:
            if not visited[ny][nx] and board[ny][nx] == '1':
                visited[ny][nx] = visited[y][x] + 1
                Q.append([ny, nx])
pprint(visited[Y - 1][X - 1])

File: 2023/05/28/미로탐색.go
package main

import (
	"bufio"
	"fmt"
	"os"
)
var file, _ = os.Open("./input/2178.txt")
var bufin = bufio.NewReader(file)
// var bufin = bufio.NewReader(os.Stdin)
var bufout = bufio.NewWriter(os.Stdout)

type P struct {
	y, x int
}
var di = [...]P{{-1, 0}, {0, 1}, {1, 0}, {0, -1}}

func main() {
	var Y, X int
	fmt.Fscan(bufin, &Y, &X)
	visited := make([][]int, Y)
	board := make([][]int, Y)
	for y := 0; y < Y; y++ {
		visited[y] = make([]int, X)
		board[y] = make([]int, X)
		var temp string
		fmt.Fscan(bufin, &temp)
		for x, v := range temp {
			board[y][x] = int(v - '0')
		}
	}
	Q :=  make([]P, 0, Y * X)
	Q = append(Q, P{0, 0})
	visited[0][0] = 1
	for len(Q) > 0 {
		p := Q[0]
		Q = Q[1:]
		for _, d := range di {
			ny := p.y + d.y
			nx := p.x + d.x
			if 0 <= ny && ny < Y && 0 <= nx && nx < X {
			if (visited[ny][nx] == 0) && (board[ny][nx] == 1) {
				Q = append(Q, P{ny, nx})
				visited[ny][nx] = visited[p.y][p.x] + 1
			}
		}
		}
	}
	for _, v := range visited {
		fmt.Println(v)
	}
	fmt.Println(visited[Y-1][X-1])
}